File: Part 2 - Regression/Section 6 - Polynomial Regression/Evans_polynomial_regression.R
# Data Preprocessing Template

# Importing the dataset

dataset = read.csv('Position_Salaries.csv')
dataset = dataset[, 2:3] # Second column inclusive.

# Splitting the dataset into the Training set and Test set
# install.packages('caTools') #Commented out as you only need to install it once.

library(caTools)

# Feature Scaling
# training_set[, 2:3] = scale(training_set[,2:3])
# test_set[, 2:3] = scale(test_set[,2:3])
regressor = lm(formula = Salary ~ .,  # Specify the relationship being examined
               data = dataset)                 # The data that will be used for learning.

# Fit polynomial regression model to the dataset.

# Create a new column
dataset$Level2 = dataset$Level^2
dataset$Level3 = dataset$Level^3
dataset$Level4 = dataset$Level^4

poly_reg = lm(formula = Salary ~ .,
              data = dataset)


# Visuallize the regressor.
library(ggplot2)

ggplot() +
  geom_point(aes(x = dataset$Level, y = dataset$Salary), # show the data as a set of points.
            colour = 'red') +
  geom_line(aes(x = dataset$Level, y = predict(poly_reg, new_data = dataset)),
            colour = 'blue') +
  ggtitle('Truth or Bluff (Polynomial Regression)') + 
  xlab('Level') +
  ylab('Salary')

# Predict a value for a particular data point
# Note how, when making the dataframe, we still need to specify a 'Level'  column.
y_pred = predict(poly_reg, data.frame(Level = 6.5,
                                       Level2 = 6.5^2,
                                       Level3 = 6.5^3,
                                       Level4 = 6.5^4))

File: Part 2 - Regression/Section 6 - Polynomial Regression/Evans_polynomial_regression.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""
Created on Sun Jan 14 15:39:08 2018

@author: evan
"""

# Data Preprocessing

# Importing the libraries

import numpy
import matplotlib.pyplot as plt
import pandas  #Useful for importing datasets.


print("Hello World")


# Importing the dataset.

dataset = pandas.read_csv('Position_Salaries.csv')

# Not the notation here: we pick our X as '1:2' rather than '1', even though these grab the same
# data (the end column is exclusive), because using this notation forces X to be treated as
# a matrix (of one column), rather than a vector.
X = dataset.iloc[:, 1:2].values
y = dataset.iloc[:, 2].values


# The library that handles linear regressions automatically handles feature scaling, so we don't need
# to worry about it.
 
# Comparing linear regression to polynomial regression.

from sklearn.linear_model import LinearRegression

linReg = LinearRegression()
linReg.fit(X, y)


from sklearn.preprocessing import PolynomialFeatures
poly_reg = PolynomialFeatures(degree = 4)
X_poly = poly_reg.fit_transform(X)

# Recall that polynomial regression models can still simultaneously be linear
# Linear refers to the coefficients, not the variables
linReg2 = LinearRegression()
linReg2.fit(X_poly, y)

#Compare regressions by visualizing the results.
# Simple linear regression
plt.scatter(X, y, color = 'red')
plt.plot(X, linReg.predict(X), color = 'blue')
plt.title('Truth or Bluff (Simple Linear Regression)')
plt.xlabel('Position Level')
plt.ylabel('Salary')
plt.show()

# Polynomial regression
X_grid = numpy.arange(min(X), max(X), 0.1)
X_gridReshaped = X_grid.reshape(((len(X_grid), 1)))

plt.scatter(X, y, color = 'red')
plt.plot(X_grid, linReg2.predict(poly_reg.fit_transform(X_gridReshaped)), color = 'blue')
plt.title('Truth or Bluff (Polynomial Regression)')
plt.xlabel('Position Level')
plt.ylabel('Salary')
plt.show()


# An employee is 'level 6.5'. Predict his salary.
# Predict a result given a value using the linear model
linReg.predict(6.5)

# Make the same prediction
linReg2.predict(poly_reg.fit_transform(6.5))



